README
Classic::Perl, version 0.05

This reinstates features deleted in recent versions of perl.

The features we currently provide are:

• split to @_ in void and scalar context. This feature was removed in
  perl 5.11. With this module, you can have it back.

• $* - This was removed in perl 5.9.

• $[ - This was removed in perl 5.15.3 and reinstated in 5.15.5. This
  may be removed from  Classic::Perl  (and then reinstated if  it  is
  again removed from Perl).


RECENT CHANGES

0.05
----
• Fixed tests to work in 5.8 and 5.6

0.04
----
• Fixed a crashing bug that occurred with ||= in 5.9 and higher.
• Fixed the changes file;  I somehow got one from a different  branch
  in the 0.03 release, listing changes that never happened.

0.03
----
• New $[ feature for 5.15+ (Thanks to Zefram for the patch!)


INSTALLATION

The easiest way to install this module is to use the CPAN  module  or
the cpan script:

    perl -MCPAN -e "install Classic::Perl"

    cpan Classic::Perl

Or you can use the following:

    perl Makefile.PL
    make
    make test
    [sudo] make install


DEPENDENCIES

This module requires perl 5 or higher.

In Perl 5.15, it requires Array::Base and String::Base.


DOCUMENTATION

After installing,  you can find documentation for these modules  with
the perldoc command.

    perldoc Classic::Perl

Or try using man (it’s faster, in my experience):

    man Classic::Perl


COPYRIGHT AND LICENCE

Copyright © 2010-12, Father Chrysostomos

This program is free software; you may redistribute it, modify it, or
both, under the same terms as perl.


Changes
Revision history for Classic::Perl

0.05    12 February, 2012
      • Fixed tests to work in 5.6 and 5.8 (Zefram)

0.04    25 September, 2011
      • Fixed a crashing bug that occurred with ||= in 5.9 and higher.
        In  5.9,  it required a constant on the right  to  cause  the
        crash. In 5.15.3 ||= would always crash.
      • Fixed the  changes  file;  I  somehow  got  one  from  a  dif-
        ferent branch in  the  0.03  release,  listing  changes  that
        never happened.

0.03    18 September, 2011
      • New $[ feature for 5.15+ (Thanks to Zefram for the patch!)

0.02a   17 October, 2010
        Make a test more resilient when certain OSes scramble line
        numbers in string evals.

0.02    10 October, 2010
      • $* in perl 5.9 and higher
      • use Classic::::Perl 5.008 import syntax
      • Fixed to work with perls built with -DDEBUGGING.

0.01    23 May, 2010
        First version, released on an unsuspecting world. It provides
        split  to  @_  in void  and  scalar  context  for  perl  5.12
        and higher.


Makefile.PL
BEGIN{ require 5; } # use

use ExtUtils::MakeMaker;

open FH, ">Perl.xs" and close FH;


my $dist = 'Classic-Perl';

my %META = (
 configure_requires => {
  'ExtUtils::MakeMaker' => 0,
 },
 build_requires => {
  'ExtUtils::MakeMaker' => 0,
  'Test::More'          => 0,
 },
 resources => {
  bugtracker => "http://rt.cpan.org/NoAuth/ReportBug.html?Queue=$dist",
  license    => 'http://dev.perl.org/licenses/',
 },
 requires => {
         $]>5.0089999 ? (
          'XSLoader' => 0,
         ) : (),
         $]>5.0130069 ? (
           re        => 0,
         ) : (),
         $]>5.0150029 ? (
           'Array::Base' => "0.004",
           'String::Base' => 0,
         ) : (),
 }
);

WriteMakefile(
    NAME             => 'Classic::Perl',
    AUTHOR           => 'Father Chrysostomos <sprout at, er, cpan.org>',
    LICENSE          => 'perl',
    VERSION_FROM     => 'lib/Classic/Perl.pm',
#    VERSION          => '0.02a',
#    XS_VERSION       => '0.02',
    ABSTRACT_FROM    => 'lib/Classic/Perl.pm',
    PREREQ_PM        => { map %{$META{$_}}, 'requires','build_requires' },
    MIN_PERL_VERSION => 5,
    META_ADD         => \%META,
    dist             => {
        COMPRESS => 'gzip -9f', SUFFIX => 'gz',
        TAR => 'COPYFILE_DISABLE=1 tar',
    },
    clean            => {
        FILES => "$dist-* *.gcov *.gcda *.gcno cover_db Perl.xs"
    },
);

unlink 'Perl.xs';

use File'Spec'Functions;
sub MY'postamble {
  my $xs_file = catfile("xs", $] < 5.0089999 ? 'old.xs' : 'new.xs');
  <<END;
Perl.xs : $xs_file
	\$(PERLRUN) -MFile::Copy "-ecopy '$xs_file', 'Perl.xs'"
END
}



lib/Classic/Perl.pm
package Classic::Perl;

my %features = map +($_ => undef)  =>=>  qw< $[ split $* >;

sub import{
 shift;
 for(@_) {
  die
    "$_ is not a feature Classic::Perl knows about at "
    . join(" line ", (caller)[1,2]) . ".\n"
   unless exists$features{$_};
  next if $] < 5.0089999;
  $_ eq '$*' and &_enable_multiline;
  next if $] < 5.0109999;
  $_ eq 'split' and $^H{Classic_Perl__split} = 1;
  next if $] < 5.0150029;
  $_ eq '$[' and $^H{'Classic_Perl__$['} = 0;
 }
 return if @_;
 return if $] < 5.0089999;
 &_enable_multiline;
 return if $] < 5.0109999;
 $^H{Classic_Perl__split} = 1;
 return if $] < 5.0150029;
 $^H{'Classic_Perl__$['} = 0;
 return;
}
sub _enable_multiline {
   $^H{'Classic_Perl__$*'} = 0,

   # It’s the autovivification of the ** glob that warns, so this is how we
   # have to suppress it. It only warns if it is created for the sake of
   # the $* variable, so ‘no warnings’ is not needed.
   *{"*"};
}
sub unimport {
 shift;
 for(@_) {
  die
    "$_ is not a feature Classic::Perl knows about at "
    . join(" line ", (caller)[1,2]) . ".\n"
   unless exists $features{$_};
  delete $^H{"Classic_Perl__$_"};
 }
 return if @_;
 if($^H{'Classic_Perl__$['}) {
  Array::Base->unimport;
  String::Base->unimport;
 }
 if(exists $^H{'Classic_Perl__$*'} and $] > 5.0130069 and $INC{"re.pm"}) {
  unimport re:: "/m";
 }
 delete @^H{map "Classic_Perl__$_", keys %features};
 return;
}

BEGIN {
 $VERSION='0.05';
 if($]>5.0089999){
  require XSLoader;
  XSLoader::load(__PACKAGE__, $VERSION);
 }
}

package Classic::::Perl;

$INC{"Classic/Perl.pm"} = $INC{"Classic//Perl.pm"} = __FILE__;

sub VERSION {
 my @features;
 push @features, '$*'    if $_[1] < 5.0089999;
 push @features, 'split' if $_[1] < 5.0109999;
 push @features, '$['    if $_[1] < 5.0150029;
 Classic::Perl->import(@features) if @features;
}

__THE__=>__END__

=head1 NAME

Classic::Perl - Selectively reinstate deleted Perl features

=head1 VERSION

Version 0.05

=head1 SYNOPSIS

  use Classic::Perl;
  # or
  use Classic::Perl 'split';

  split //, "smat";
  print join " ", @_; # prints "s m a t"

  no Classic::Perl;
  @_ = ();
  split //, "smat";
  print join " ", @_;
    # prints "s m a t" in perl 5.10.x; nothing in 5.12

  use Classic::Perl '$[';
  $[ = 1;
  print qw(a b c d)[2]; # prints "b"

  use Classic::Perl '$*';
  $* = 1;
  print "yes\n" if "foo\nbar" =~ /^bar/; # prints yes

=head1 DESCRIPTION

Classic::Perl restores some Perl features that have been deleted in the
latest versions. By 'classic' we mean as of perl 5.8.x.

The whole idea is that you can put C<use Classic::Perl> at the top of an
old script or module (or a new one, if you like the features that are out
of vogue) and have it continue to work.

In versions of perl prior to 5.10, this module simply does nothing.

=head1 ENABLING FEATURES

To enable all features, simply use C<use Classic::Perl;>. To disable
whatever Classic::Perl enabled, write C<no Classic::Perl;>. These are
lexically-scoped, so:

  {
     use Classic::Perl;
     # ... features on here ...
  }
  # ... features off here ...

To enable or disable a specific set of features, pass them as arguments to
C<use> or C<no>:

  use Classic::Perl qw< $[ split $* >;

To enable features that still existed in a given version of perl, put
I<four> colons in your C<use> statement, followed by the perl version. Only
plain numbers (C<5.008>) are currently supported. Don't use v-strings
(C<v5.8.0>).

  use Classic::::Perl 5.016; # does nothing (yet)
  use Classic::::Perl 5.014; # enables $[, but not split or $*
  use Classic::::Perl 5.010; # enables $[ and split, but not $*
  use Classic::::Perl 5.008; # enables everything

This is not guaranteed to do anything reasonable if used with C<no>.

=head1 THE FEATURES THEMSELVES

=over

=item $[

This feature provides the C<$[> variable, which, when set to an integer
other than zero, offsets indices into arrays and strings.  For example,
setting it to 1 (almost the only non-zero value actually used) means
that the first element in an array has index 1 rather than the usual 0.
The index offset is lexically scoped, as C<$[> has been as of Perl 5.10,
unlike its behaviour in Perl 5.0-5.8 (file-scoped) and Perl 1-4 (global).

This is deprecated in Perl, but has not yet been removed.  If it is
removed, Classic::Perl will continue to provide it.

=item split

This features provides C<split> to C<@_> in void and scalar context.

This was removed from perl in 5.11.

=item $*

This feature provides the C<$*> variable, which, when set to an integer
other than zero, puts an implicit C</m> on every regular expression.

Unlike the C<$*> variable in perl 5.8 and earlier, this only works at
compile-time and is lexically
scoped (like C<$[> in 5.10-5.14). It only works with constant values.
C<$* = $val> does not work.

<$*> was removed in perl 5.9.

=back

=head1 BUGS

Please report any bugs you find via L<http://rt.cpan.org> or
L<bug-Classic-Perl@rt.cpan.org>.

=head1 ACKNOWLEDGEMENTS

Much of the structural code in the XS file was stolen from Vincent Pit's
C<autovivification> module and tweaked. The F<ptable.h> file was taken
straight from his module without modifications. (I have been subsequently
informed that he stole it from B::Hooks::OP::Check, which pilfered it from
autobox, which filched it from perl. :-)

Andrew Main (Zefram) added support for C<$[> in 5.16.

=head1 SINE QUIBUS NON

L<perl> 5 or higher

In Perl 5.16 and higher, L<Array::Base> and L<String::Base> are required.

=head1 COPYRIGHT

Copyright (C) 2010-12 Father Chrysostomos

  use Classic'Perl;
  split / /, 'org . cpan @ sprout';
  print reverse "\n", @_;

This program is free software; you may redistribute it, modify it or both
under the same terms as perl.

=head1 SEE ALSO

L<Array::Base>, L<String:Base>,
L<perl>, L<C<split> in perlfunc|perlfunc/split>,
L<C<$*> in perlvar|perlvar/$*>,
C<C<$[> in perlvar|perlvar/$[>

L<any::feature> is an experimental module that backports new Perl features
to older versions.

The L<Modern::Perl> module enables various pragmata which are currently
popular.
